cmd/s3-peer-client.go

```diff
@@ -19,125 +19,74 @@ package cmd
 import (
 	"encoding/json"
 	"fmt"
-	"net/rpc"
 	"net/url"
 	"path"
 	"sync"
-	"time"
 )
 
 type s3Peers struct {
-	// A map of peer server address (in `host:port` format) to RPC
-	// client connections.
-	rpcClients map[string]*AuthRPCClient
+	// A map of peer server address (in `host:port` format) to
+	// BucketMetaState interface
+	bmsClients map[string]BucketMetaState
 
-	mutex *sync.RWMutex
-
-	// Is single-node?
-	isDistXL bool
+	// Is distributed?
+	IsDistXL bool
 
 	// Slice of all peer addresses (in `host:port` format).
-	peers []string
+	Peers []string
 }
 
-func initGlobalS3Peers(eps []*url.URL) {
-	// Get list of de-duplicated peers.
-	peers := getAllPeers(eps)
+func mkPeersFromEndpoints(eps []*url.URL) s3Peers {
+	bmsc := make(map[string]BucketMetaState)
 
-	// Initialize global state.
-	globalS3Peers = s3Peers{
-		rpcClients: make(map[string]*AuthRPCClient),
-		mutex:      &sync.RWMutex{},
-	}
+	// Setup local server state change handler
+	bmsc[globalMinioAddr] = &localBMS{ObjectAPI: newObjectLayerFn}
 
-	// Initialize each peer connection.
-	for _, peer := range peers {
-		globalS3Peers.InitS3PeerClient(peer)
-	}
+	peers := []string{globalMinioAddr}
 
-	// Save new peers
-	globalS3Peers.peers = peers
-
-	// store if this is a distributed setup or not.
-	globalS3Peers.isDistXL = len(globalS3Peers.peers) > 1
-}
+	for _, ep := range eps {
+		if ep == nil || ep.Host == "" || ep.Host == globalMinioAddr {
+			continue
+		}
 
-func (s3p *s3Peers) GetPeers() []string {
-	return s3p.peers
-}
+		// Check if the remote host has been added already
+		if _, ok := bmsc[ep.Host]; !ok {
+			cfg := authConfig{
+				accessKey:   serverConfig.GetCredential().AccessKeyID,
+				secretKey:   serverConfig.GetCredential().SecretAccessKey,
+				address:     ep.Host,
+				secureConn:  isSSL(),
+				path:        path.Join(reservedBucket, s3Path),
+				loginMethod: "S3.LoginHandler",
+			}
 
-func (s3p *s3Peers) GetPeerClient(peer string) *AuthRPCClient {
-	// Take a read lock
-	s3p.mutex.RLock()
-	defer s3p.mutex.RUnlock()
-	return s3p.rpcClients[peer]
-}
+			bmsc[ep.Host] = &remoteBMS{newAuthClient(&cfg)}
 
-// Initializes a new RPC connection (or closes and re-opens if it
-// already exists) to a peer. Note that peer address is in `host:port`
-// format.
-func (s3p *s3Peers) InitS3PeerClient(peer string) {
-	// Take a write lock
-	s3p.mutex.Lock()
-	defer s3p.mutex.Unlock()
-
-	if s3p.rpcClients[peer] != nil {
-		_ = s3p.rpcClients[peer].Close()
-		delete(s3p.rpcClients, peer)
+			peers = append(peers, ep.Host)
+		}
 	}
-	authCfg := &authConfig{
-		accessKey:   serverConfig.GetCredential().AccessKeyID,
-		secretKey:   serverConfig.GetCredential().SecretAccessKey,
-		address:     peer,
-		secureConn:  isSSL(),
-		path:        path.Join(reservedBucket, s3Path),
-		loginMethod: "S3.LoginHandler",
+
+	return s3Peers{
+		bmsClients: bmsc,
+		IsDistXL:   len(peers) > 1,
+		Peers:      peers,
```

Comment:
There is trade-off I discovered while re-working this patch. `GetPeerClient` is used in a frequently hit code path in [`notify-listener.go`](https://github.com/minio/minio/blob/master/cmd/notify-listener.go#L61) to send event notifications to peers connected to listeners. If we do not use a map, we will need to iterate over a 16-length slice each time to locate the client object to send the notification - if we store a map, we do a fast(er) map lookup.

I'm of the opinion that we use the map since this usage of (a small amount of) memory is helping performance. What say?